### ISSUE TITLE
CalibreDB sessions is not recovered from error

### ISSUE THREAD

**Issue by marcinkoziej** [janeczku/calibre-web#1910]
**Describe the bug/problem**
A clear and concise description of what the bug is. If you are asking for support, please check our [Wiki](https://github.com/janeczku/calibre-web/wiki) if your question is already answered there.

Hello. After execution of statement in Calibre DB, the `calibre_db.session` becomes invalid, and the app is unusable until it is restarted.

For example, it can be an error like this: `*** sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: authors.name`.

Because session is held in global object `calibre_db`, any further use will result in following error:

`*** sqlalchemy.exc.InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) UNIQUE constraint failed: authors.name`

**To Reproduce**
1. Insert a `ipdb` breakpoint somewhere in routes, so you can access REPL 
1. Start the app `python cps.py -p app.db`
2. Open the app in browser
3. The PDB should break the app. 
4. Generate an error:
```
calibre_db.session.add(db.Authors("pe","pe","pe"))
calibre_db.session.add(db.Authors("pe","pe","pe"))
calibre_db.session.commit()
```
it should result in:
```
*** sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: authors.name
[SQL: INSERT INTO authors (name, sort, link) VALUES (?, ?, ?)]
[parameters: ('pe', 'pe', 'pe')]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
```
5. write `cont` to leave the debugger.
6. now try to refresh app in browser, it will throw an error (if you see an error about "entries" not being set, it's just because of session exception)

**Environment (please complete the following information):**
 - OS: Ubuntu 20.10
 - Python version: python 3.8
 - Calibre-Web version: 0.6.11
 - Docker container: LinuxServer
 - Browser: Firefox latest
 - 
**Additional context**
I do know that in many places there is exception handling where session.rollback() is called. However, perhaps not all errors are caught.
I have made a workaround to check whether the session is active and re-init it if necessary, but this is not an elegant solution.
I think a better one would involve a global error handler which would re-initialize the global session, or to use one session object per request.

Code of my workaround is:
```
diff --git a/cps/db.py b/cps/db.py
index c08a3ea4..60f283f2 100644
--- a/cps/db.py
+++ b/cps/db.py
@@ -433,7 +433,7 @@ class CalibreDB():
     def __init__(self, expire_on_commit=True):
         """ Initialize a new CalibreDB session
         """
-        self.session = None
+        self._session = None
         if self._init:
             self.initSession(expire_on_commit)
 
@@ -441,10 +441,17 @@ class CalibreDB():
 
 
     def initSession(self, expire_on_commit=True):
-        self.session = self.session_factory()
-        self.session.expire_on_commit = expire_on_commit
+        self._session = self.session_factory()
+        self._session.expire_on_commit = expire_on_commit
         self.update_title_sort(self.config)
 
+    @property
+    def session(self):
+        if not self._session.is_active:
+            self._session.rollback()
+            self.initSession()
+        return self._session
+
     @classmethod
     def setup_db(cls, config, app_db_path):
         cls.config = config
```

**Comment by marcinkoziej** [comment]
Upon further investigation, I found that I had 

```
58|Bruce Fink, Richard Feldstein, Maire Jaanus|Bruce Fink, Richard Feldstein, Maire Jaanus|
118|Bruce Fink| Richard Feldstein| Maire Jaanus|Bruce Fink, Richard Feldstein, Maire Jaanus|
```
and Calibre tried to do update the second row to same name as first row:
```
[SQL: UPDATE authors SET name=? WHERE authors.id = ?]
[parameters: ('Bruce Fink, Richard Feldstein, Maire Jaanus', 118)]
```

Still, this error has "stuck" somehow in the session, preventing even opening pages - it seems that because of no auto-flushing, it remained in uncommited way and was forced to commit by 
`    entries = calibre_db.session.query(db.Authors, func.count('books_authors_link.book').label('count'))`
in ` "/app/calibre-web/cps/web.py", line 796, in author_list`

**Comment by OzzieIsaacs** [comment]
Yeah there is a problem I'm aware of it. Your are the 2nd or 3rd person having trouble with duplicate authors. I tried to change it to autocommit/session per request, but this caused a lot of other problems at it wasn't even running anymore (as far as I remember). The remains of the experiment are in the session branch
